examples/control-vector-generator/pca.hpp

```diff
@@ -0,0 +1,253 @@
+        ggml_set_name(square, "square");
+        //model.square = ggml_scale_inplace(ctx0, model.square, 0.0);
+    }
+
+    struct ggml_tensor * b_tensor;
+
+    for (int i = 0; i < nb_iterations; ++i) {
+        // b_tensor = square * eigenvector^T
+        b_tensor = ggml_mul_mat(ctx0, square, model.eigenvector);
+        ggml_set_name(b_tensor, "b_tensor");
+
+        // normalize
+        b_tensor = ggml_div_inplace(ctx0,
+            b_tensor,
+            ggml_sqrt_inplace(ctx0, ggml_sum_rows(ctx0, ggml_sqr(ctx0, b_tensor)))
+        );
+    }
+
+    // calculate distance
+    struct ggml_tensor * distance;
+    {
+        distance = ggml_sub(ctx0, model.eigenvector, b_tensor);
+        ggml_set_name(distance, "distance");
+        distance = ggml_sqrt_inplace(ctx0,
+            ggml_sum_rows(ctx0, ggml_sqr_inplace(ctx0, distance)));
+    }
+
+    // build operations nodes
+    ggml_build_forward_expand(gf, distance);
+
+    // delete the temporally context used to build the graph
+    ggml_free(ctx0);
+    return gf;
+}
+
+struct ggml_tensor * compute_piter(
+        const pca_model & model,
+        struct ggml_cgraph * gf,
+        ggml_gallocr_t allocr,
+        int n_threads) {
+    // allocate tensors
+    ggml_gallocr_alloc_graph(allocr, gf);
+
+    if (ggml_backend_is_cpu(model.backend)) {
+        ggml_backend_cpu_set_n_threads(model.backend, n_threads);
+    }
+
+#ifdef GGML_USE_METAL
+    if (ggml_backend_is_metal(model.backend)) {
+        ggml_backend_metal_set_n_cb(model.backend, n_threads);
+    }
+#endif
+
+    ggml_backend_graph_compute(model.backend, gf);
+
+    // in this case, the output tensor is the last one in the graph
+    return gf->nodes[gf->n_nodes - 1];
+}
+
+static void power_iteration(
+        struct ggml_tensor * input,
+        struct ggml_tensor * output,
+        int n_threads,
+        int maxIterations = 1000,
+        float tolerance = 1e-7) {
+    printf("in power iteration\n");
+    int n_embd = input->ne[0]; // shape of input: [n_embd, m]
+
+    pca_model model;
+    load_pca_model(model, input);
+
+    ggml_gallocr_t allocr = NULL;
+
+    struct ggml_init_params host_params = {
+        /*.mem_size   =*/ (n_embd * sizeof(float) + ggml_tensor_overhead()) * 4u,
+        /*.mem_buffer =*/ NULL,
+        /*.no_alloc   =*/ false,
+    };
+    struct ggml_context * host_ctx = ggml_init(host_params);
+
+    struct ggml_tensor * host_old_eigenvector = ggml_new_tensor_1d(host_ctx, GGML_TYPE_F32, n_embd);
+    struct ggml_tensor * host_new_eigenvector = ggml_new_tensor_1d(host_ctx, GGML_TYPE_F32, n_embd);
+
+    for (int iter = 0; iter < maxIterations; ++iter) {
+        if (allocr) {
+            ggml_gallocr_free(allocr);
+        }
+        allocr = ggml_gallocr_new(ggml_backend_get_default_buffer_type(model.backend));
+        struct ggml_cgraph * gf = build_graph_piter(model, iter == 0);
+        ggml_graph_dump_dot(gf, nullptr, "/tmp/_cgraph.dot");
```

Comment:
@christianazinn You can use this command to dump the cgraph to graphviz dot format, then use https://dreampuf.github.io/GraphvizOnline/ to convert it to svg file